tests/src/test/java/org/eclipse/hono/tests/client/ClientTestBase.java

```diff
@@ -230,7 +201,58 @@ private void disconnect(final TestContext ctx) {
      * @param messageConsumer The handler to invoke for every message received.
      * @return A future succeeding with the created consumer.
      */
-    abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
+    protected abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
+
+    /**
+     * Connects the downstream consumer client for connection to the messaging network.
+     * 
+     * @return The client for connecting to the messaging network.
+     */
+    protected abstract Future<HonoClient> connectDownstreamClient(ProtonClientOptions options);
+
+    /**
+     * Connects the AMQP 1.0 client to one of Hono's services.
+     * 
+     * @param options The options to use for the connection.
+     * 
+     * @return The client for connecting to a Hono service.
+     */
+    protected abstract Future<HonoClient> connectToService(ProtonClientOptions options, TestContext ctx);
+
+    /**
+     * Sends a message to the downstream consumer through a Hono service.
+     * 
+     * @param deviceId The id of the device sending the message.
+     * @param ctx The Vert.x test context.
+     */
+    protected abstract void sendMessage(String payload, TestContext ctx);
+
+    /**
+     * The id of the tenant under test.
+     * 
+     * @return The tenant identifier created during the test.
+     */
+    protected abstract String getTenantId();
+
+    /**
+     * Gets the random device-id that was generated during a test execution.
+     * 
+     * @return The random device-id.
+     */
+    protected abstract String getRandomDevice();
```

Comment:
same here

Reply:
please note that there is a `@Test` in the `ClientTestBase` that needs access to the device that was generated during connection to the adapter. for hono-messaging, we can generate a new one but not for the adapter(since a connection was already established before the test starts to execute)